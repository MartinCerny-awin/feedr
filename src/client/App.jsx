import React, { useEffect, useMemo, useState } from 'react';

import MenuSummary from './components/MenuSummary';
import MenuPreview from './components/MenuPreview';
import Filters from './components/Filters';

import './App.css';

const App = () => {
  const [searchTerm, setSearchTerm] = useState('');
  const [items, setItems] = useState([]);
  const [selectedIds, setSelectedIds] = useState([]);

  useEffect(() => {
    (async () => {
      const response = await fetch('/api/items');
      const json = await response.json();

      setItems(json.items);
    })();
  }, []);

  const handleAddItem = (id) => {
    if (!selectedIds.find((selectedId) => selectedId === id)) {
      setSelectedIds([...selectedIds, id]);
    }
  };

  const handleRemoveItem = (id) => {
    setSelectedIds(selectedIds.filter((selectedId) => selectedId !== id));
  };

  const filteredItems = useMemo(() => {
    if (!searchTerm) {
      return items;
    }
    return items.filter(({ name }) =>
      RegExp(searchTerm.toLowerCase()).test(name.toLowerCase()),
    );
  }, [items, searchTerm]);

  const selectedItems = useMemo(
    () => items.filter((item) => selectedIds.includes(item.id)),
    [items, selectedIds],
  );

  const selectedDietaries = useMemo(() => {
    const calculateDietaries = {};

    selectedItems.forEach(({ dietaries }) => {
      dietaries.forEach((dietary) => {
        calculateDietaries[dietary] = Object.prototype.hasOwnProperty.call(
          calculateDietaries,
          dietary,
        )
          ? calculateDietaries[dietary] + 1
          : 1;
      });
    });
    return calculateDietaries;
  }, [selectedItems, selectedIds]);

  return (
    <div className="wrapper">
      <MenuSummary
        numberOfSelectedItems={selectedIds.length}
        selectedDietaries={selectedDietaries}
      />
      <div className="container menu-builder">
        <div className="row">
          <div className="col-4">
            <Filters
              filteredItems={filteredItems}
              handleAddItem={handleAddItem}
              searchTerm={searchTerm}
              setSearchTerm={setSearchTerm}
            />
          </div>
          <div className="col-8">
            <MenuPreview
              selectedItems={selectedItems}
              handleRemoveItem={handleRemoveItem}
            />
          </div>
        </div>
      </div>
    </div>
  );
};

export default App;
