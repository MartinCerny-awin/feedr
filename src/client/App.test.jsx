import React from 'react';
import { act, render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import App from './App';

global.fetch = jest.fn();
global.fetch.mockImplementation(() => ({
  ok: true,
  json: () => ({
    items: [
      {
        id: 1001,
        name: 'Test1',
        dietaries: ['dietary-a', 'dietary-b'],
      },
      {
        id: 1002,
        name: 'This is name!',
        dietaries: ['dietary-b', 'dietary-c'],
      },
      {
        id: 1003,
        name: 'Test3',
        dietaries: ['dietary-a'],
      },
    ],
  }),
}));

let container;
describe('App', () => {
  beforeEach(async () => {
    await act(async () => {
      ({ container } = await render(<App />));
    });
  });

  it('renders items for selection', () => {
    expect(container).toMatchSnapshot();
  });

  it('searches for items', () => {
    const input = screen.getByPlaceholderText('Name');

    fireEvent.change(input, { target: { value: 'Name' } });

    expect(screen.queryByText('Test1')).not.toBeInTheDocument();
    expect(screen.getByText('This is name!')).toBeInTheDocument();
    expect(screen.queryByText('Test3')).not.toBeInTheDocument();
  });

  it('selects item, show number of items, show selected dietaries', () => {
    expect(screen.queryAllByText('Test1')).toHaveLength(1);

    userEvent.click(screen.getByText('Test1'));

    expect(container).toMatchSnapshot();
    expect(screen.getByText('1 items')).toBeInTheDocument();
    expect(screen.queryAllByText('dietary-a')).toHaveLength(4);
    expect(screen.queryAllByText('Test1')).toHaveLength(2);
  });

  it('select multiple items and removes item', () => {
    userEvent.click(screen.getByText('Test1'));
    userEvent.click(screen.getByText('Test3'));
    userEvent.click(screen.queryAllByText('x')[0]);

    expect(container).toMatchSnapshot();
    expect(screen.queryAllByText('Test1')).toHaveLength(1);
  });

  it('does not add the same item twice', () => {
    userEvent.click(screen.getByText('Test1'));
    userEvent.click(screen.queryAllByText('Test1')[0]);

    expect(container).toMatchSnapshot();
    expect(screen.queryAllByText('Test1')).toHaveLength(2);
  });
});
