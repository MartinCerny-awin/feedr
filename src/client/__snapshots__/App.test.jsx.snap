// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App does not add the same item twice 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="menu-summary"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-6 menu-summary-left"
          >
            <span>
              1
               items
            </span>
          </div>
          <div
            class="col-6 menu-summary-right"
          >
            1
            x 
            <span
              class="dietary"
            >
              dietary-a
            </span>
            1
            x 
            <span
              class="dietary"
            >
              dietary-b
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container menu-builder"
    >
      <div
        class="row"
      >
        <div
          class="col-4"
        >
          <div
            class="filters"
          >
            <input
              class="form-control"
              placeholder="Name"
              value=""
            />
          </div>
          <ul
            class="item-picker"
          >
            <li
              class="item"
            >
              <h2>
                Test1
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-a
                </span>
                <span
                  class="dietary"
                >
                  dietary-b
                </span>
              </p>
            </li>
            <li
              class="item"
            >
              <h2>
                This is name!
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-b
                </span>
                <span
                  class="dietary"
                >
                  dietary-c
                </span>
              </p>
            </li>
            <li
              class="item"
            >
              <h2>
                Test3
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-a
                </span>
              </p>
            </li>
          </ul>
        </div>
        <div
          class="col-8"
        >
          <h2>
            Menu preview
          </h2>
          <ul
            class="menu-preview"
          >
            <li
              class="item"
            >
              <h2>
                Test1
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-a
                </span>
                <span
                  class="dietary"
                >
                  dietary-b
                </span>
              </p>
              <button
                class="remove-item"
                type="button"
              >
                x
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`App renders items for selection 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="menu-summary"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-6 menu-summary-left"
          >
            <span>
              0
               items
            </span>
          </div>
          <div
            class="col-6 menu-summary-right"
          />
        </div>
      </div>
    </div>
    <div
      class="container menu-builder"
    >
      <div
        class="row"
      >
        <div
          class="col-4"
        >
          <div
            class="filters"
          >
            <input
              class="form-control"
              placeholder="Name"
              value=""
            />
          </div>
          <ul
            class="item-picker"
          >
            <li
              class="item"
            >
              <h2>
                Test1
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-a
                </span>
                <span
                  class="dietary"
                >
                  dietary-b
                </span>
              </p>
            </li>
            <li
              class="item"
            >
              <h2>
                This is name!
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-b
                </span>
                <span
                  class="dietary"
                >
                  dietary-c
                </span>
              </p>
            </li>
            <li
              class="item"
            >
              <h2>
                Test3
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-a
                </span>
              </p>
            </li>
          </ul>
        </div>
        <div
          class="col-8"
        >
          <h2>
            Menu preview
          </h2>
          <ul
            class="menu-preview"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`App select multiple items and removes item 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="menu-summary"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-6 menu-summary-left"
          >
            <span>
              1
               items
            </span>
          </div>
          <div
            class="col-6 menu-summary-right"
          >
            1
            x 
            <span
              class="dietary"
            >
              dietary-a
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container menu-builder"
    >
      <div
        class="row"
      >
        <div
          class="col-4"
        >
          <div
            class="filters"
          >
            <input
              class="form-control"
              placeholder="Name"
              value=""
            />
          </div>
          <ul
            class="item-picker"
          >
            <li
              class="item"
            >
              <h2>
                Test1
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-a
                </span>
                <span
                  class="dietary"
                >
                  dietary-b
                </span>
              </p>
            </li>
            <li
              class="item"
            >
              <h2>
                This is name!
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-b
                </span>
                <span
                  class="dietary"
                >
                  dietary-c
                </span>
              </p>
            </li>
            <li
              class="item"
            >
              <h2>
                Test3
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-a
                </span>
              </p>
            </li>
          </ul>
        </div>
        <div
          class="col-8"
        >
          <h2>
            Menu preview
          </h2>
          <ul
            class="menu-preview"
          >
            <li
              class="item"
            >
              <h2>
                Test3
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-a
                </span>
              </p>
              <button
                class="remove-item"
                type="button"
              >
                x
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`App selects item, show number of items, show selected dietaries 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="menu-summary"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-6 menu-summary-left"
          >
            <span>
              1
               items
            </span>
          </div>
          <div
            class="col-6 menu-summary-right"
          >
            1
            x 
            <span
              class="dietary"
            >
              dietary-a
            </span>
            1
            x 
            <span
              class="dietary"
            >
              dietary-b
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container menu-builder"
    >
      <div
        class="row"
      >
        <div
          class="col-4"
        >
          <div
            class="filters"
          >
            <input
              class="form-control"
              placeholder="Name"
              value=""
            />
          </div>
          <ul
            class="item-picker"
          >
            <li
              class="item"
            >
              <h2>
                Test1
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-a
                </span>
                <span
                  class="dietary"
                >
                  dietary-b
                </span>
              </p>
            </li>
            <li
              class="item"
            >
              <h2>
                This is name!
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-b
                </span>
                <span
                  class="dietary"
                >
                  dietary-c
                </span>
              </p>
            </li>
            <li
              class="item"
            >
              <h2>
                Test3
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-a
                </span>
              </p>
            </li>
          </ul>
        </div>
        <div
          class="col-8"
        >
          <h2>
            Menu preview
          </h2>
          <ul
            class="menu-preview"
          >
            <li
              class="item"
            >
              <h2>
                Test1
              </h2>
              <p>
                <span
                  class="dietary"
                >
                  dietary-a
                </span>
                <span
                  class="dietary"
                >
                  dietary-b
                </span>
              </p>
              <button
                class="remove-item"
                type="button"
              >
                x
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
